### PR TITLE
Issue #4164: Custom tags' content not to be parsed as HTML

### PR DESCRIPTION
**Pull request by ps-sp**
#4164 : Custom tags' content not to be parsed as HTML

### CODE REVIEW

**Comment by ps-sp** [comment]
$ cat Test.java 
  
     /**
     * standard inline tag: {@code a < b} , parsed ok.
     * standard inline tag: {@code a<b} , parsed ok.
     * Non standard {@preformat a < b} parsed ok.
     * Non standard {@preformat a<b} PREVIOUSLY parsed with error.
     */
    public class Test {
    }

java -jar checkstyle-7.8-SNAPSHOT-all.jar -J Test.java

    CLASS_DEF -> CLASS_DEF [7:0]
    |--MODIFIERS -> MODIFIERS [7:0]
    |   |--BLOCK_COMMENT_BEGIN -> /* [1:0]
    |   |   |--COMMENT_CONTENT -> *\r\n * standard inline tag: {@code a < b} , parsed ok.\r\n * standard inline tag: {@code a<b} , parsed ok.\r\n * Non standard {@preformat a < b} parsed ok.\r\n * Non standard {@preformat a<b} PREVIOUSLY parsed with error.\r\n  [1:2]
    |   |   |   `--JAVADOC -> \r\n * standard inline tag: {@code a < b} , parsed ok.\r\n * standard inline tag: {@code a<b} , parsed ok.\r\n * Non standard {@preformat a < b} parsed ok.\r\n * Non standard {@preformat a<b} PREVIOUSLY parsed with error.\r\n <EOF> [1:3]
    |   |   |       |--NEWLINE -> \r\n [1:3]
    |   |   |       |--LEADING_ASTERISK ->  * [2:0]
    |   |   |       |--TEXT ->  standard inline tag:  [2:2]
    |   |   |       |--JAVADOC_INLINE_TAG -> {@code a < b} [2:24]
    |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [2:24]
    |   |   |       |   |--CODE_LITERAL -> @code [2:25]
    |   |   |       |   |--WS ->   [2:30]
    |   |   |       |   |--TEXT -> a < b [2:31]
    |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [2:37]
    |   |   |       |--TEXT ->  , parsed ok. [2:38]
    |   |   |       |--NEWLINE -> \r\n [2:51]
    |   |   |       |--LEADING_ASTERISK ->  * [3:0]
    |   |   |       |--TEXT ->  standard inline tag:  [3:2]
    |   |   |       |--JAVADOC_INLINE_TAG -> {@code a<b} [3:24]
    |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [3:24]
    |   |   |       |   |--CODE_LITERAL -> @code [3:25]
    |   |   |       |   |--WS ->   [3:30]
    |   |   |       |   |--TEXT -> a<b [3:31]
    |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [3:35]
    |   |   |       |--TEXT ->  , parsed ok. [3:36]
    |   |   |       |--NEWLINE -> \r\n [3:49]
    |   |   |       |--LEADING_ASTERISK ->  * [4:0]
    |   |   |       |--TEXT ->  Non standard  [4:2]
    |   |   |       |--JAVADOC_INLINE_TAG -> {@preformat a < b} [4:16]
    |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [4:16]
    |   |   |       |   |--CUSTOM_NAME -> @preformat [4:17]
    |   |   |       |   |--WS ->   [4:27]
    |   |   |       |   |--DESCRIPTION -> a < b [4:28]
    |   |   |       |   |   `--TEXT -> a < b [4:28]
    |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [4:33]
    |   |   |       |--TEXT ->  parsed ok. [4:34]
    |   |   |       |--NEWLINE -> \r\n [4:45]
    |   |   |       |--LEADING_ASTERISK ->  * [5:0]
    |   |   |       |--TEXT ->  Non standard  [5:2]
    |   |   |       |--JAVADOC_INLINE_TAG -> {@preformat a<b} [5:16]
    |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [5:16]
    |   |   |       |   |--CUSTOM_NAME -> @preformat [5:17]
    |   |   |       |   |--WS ->   [5:27]
    |   |   |       |   |--DESCRIPTION -> a<b [5:28]
    |   |   |       |   |   `--TEXT -> a<b [5:28]
    |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [5:31]
    |   |   |       |--TEXT ->  PREVIOUSLY parsed with error. [5:32]
    |   |   |       |--NEWLINE -> \r\n [5:62]
    |   |   |       |--TEXT ->   [6:0]
    |   |   |       `--EOF -> <EOF> [6:1]
    |   |   `--BLOCK_COMMENT_END -> */ [6:1]
    |   `--LITERAL_PUBLIC -> public [7:0]
    |--LITERAL_CLASS -> class [7:7]
    |--IDENT -> Test [7:13]
    `--OBJBLOCK -> OBJBLOCK [7:18]
        |--LCURLY -> { [7:18]
        `--RCURLY -> } [8:0]

**Comment by rnveach** [comment]
@ps-sp Again, can we have some javadoc print tests somewhere to verify the issue is fixed.

**Comment by ps-sp** [comment]
@rnveach done

**Comment by ps-sp** [comment]
@rnveach @romani I was facing the issues with `org.apache.maven.plugins:maven-checkstyle-plugin:2.17:check` on my local ! same issues seems to persist here to ! It seems to be related to commit: [4c921655f64dce7538306fbd275234a53cc778cd](https://github.com/checkstyle/checkstyle/commit/4c921655f64dce7538306fbd275234a53cc778cd) in main/master.

**Comment by ps-sp** [comment]
@romani  @rnveach @Vladlis Another issue with sventu still persits ! I don't feel it relates to the commit this PR ! Please have a look ! Please correct me if I am wrong. Thanks

**Comment by rnveach** [comment]
>  I was facing the issue with org.apache.maven.plugins:maven-checkstyle-plugin:2.17:check on my local !

What was your issue?
CI failure is because we are not using all sevntu checks in checkstyle. `MoveVariableInsideIf` is missing.
I will make a PR with a fix.

**Comment by rnveach** [comment]
Missing check should be finished in latest master. Please rebase.

**Comment by ps-sp** [comment]
@rnveach I was referring to missing check only in my previous comment ! In the comment before that I was referring to the issue that maven was unable to resolve the sventu artifact ! It got resovled later when the pom got updated in one of the recent commits !

File: src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java [hunk 1]
```diff
@@ -5772,4 +5772,61 @@ public static ParseTree treeCustomTagWithDot()
         return javadocContext0;
     }
 
+    public static ParseTree treeInlineCustomJavadocTag()
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java]: @romani We should remove this way of validation and switch to verifying via `AstTreeStringPrinter` like we do with `verifyAst`. Reading, understanding, and modifying custom written trees is not easy.
I will make an issue for this.

File: src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java [hunk 2]
```diff
@@ -5772,4 +5772,61 @@ public static ParseTree treeCustomTagWithDot()
...
+    {
```
Comment (Vladlis) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java]: Please move left curly to the previous line

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java]: Why didn't `LeftCurlyCheck` pick this up?

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java]: `LeftCurlyCheck` isn't picking this up because we are suppressing everything for these files at https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml#L56-L57 .
I will make new issues.

Comment (ps-sp) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/ParseTreeBuilder.java]: DONE

**Comment by romani** [comment]
@ps-sp , please do not use "!" In your comments, at all. Use simple ".".

preformat tag in tree results in node DESCRIPTION , but "code" is not , please explain why.

**Comment by ps-sp** [comment]
@romani yes, I will take care not to. 

That's how the grammar is: https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4#L904

Changing that line to this: `CODE_LITERAL (WS | NEWLINE | LEADING_ASTERISK)* description?` would result in a `description` node in code tag too

**Comment by rnveach** [comment]
@ps-sp There are file conflicts, please rebase.

**Comment by ps-sp** [comment]
@rnveach rebased.

**Comment by romani** [comment]
> That's how the grammar is

will be addressed separately at #4385

**Comment by ps-sp** [comment]
//mistakenly clicked on close pull request.

**Comment by rnveach** [comment]
@romani to use our current process with a new antlr compare, we have to print out every tree line as a violation.
I made a quick check to test this out and it works, but the XML files may be larger. I don't recommend using it on large projects like openjdk.
https://github.com/rnveach/checkstyle/commits/javadoc_regression_check
The new check is `JavadocTreeCheck`. Because this check doesn't exist in `master`, it has to be temporarily added to the master and PR branch when run with `diff.groovy`.

If we want a better process, lets start a new issue on it in `contribution`.

**Comment by Vladlis** [comment]
@rnveach great idea of how we can reuse an existing tool!

@ps-sp please, start a new issue on it in contribution and lets discuss details there

**Comment by ps-sp** [comment]
[Issue](https://github.com/checkstyle/contribution/issues/237) has been created in contribution.

**Comment by ps-sp** [comment]
Why does it make more sense for javadoc to be under Modifiers rather than the type def ?

**Comment by rnveach** [comment]
Comment insertion is probably not ideal. It is inserted programmatically, outside of antlr. See [TreeWalker.appendHiddenCommentNodes](https://github.com/checkstyle/checkstyle/blob/de50d3465849b83d25910590e1f5f39a25fe6e2c/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L521)

**Comment by ps-sp** [comment]
@rnveach Sorry I didn't quite get you ? Perhaps I phrased my question wrong. Let me rephrase it, Would it make better sense if BLOCK_COMMENT_BEGIN was a child of CLASS_DEF rather than MODIFIERS ?

**Comment by rnveach** [comment]
@ps-sp It probably would make more sense.
We did not put the comment there on purpose. It was determined by the algorithm I stated above. Comments can be almost anywhere in the tree, so we use an algorithm to determine it's placement instead of predetermining where it should go. Changing where the comment is inserted requires changing that algorithm, most likely creating a specific override.

**Comment by ps-sp** [comment]
The position of insertion is determined in that method itself actually. ANTLR is just helping send the comments on hidden channels I feel and thus the hidden content is not inserted as nodes but as fields of nodes in the parse tree returned by ANTLR.

**Comment by ps-sp** [comment]
@rnveach @romani @Vladlis 
Should we make an issue for [this](https://github.com/checkstyle/checkstyle/pull/4378#issuecomment-305076860) and try to fix it ?

**Comment by ps-sp** [comment]
Also, [walk](https://github.com/checkstyle/checkstyle/blob/de50d3465849b83d25910590e1f5f39a25fe6e2c/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L180-#L184) calls and `appendHiddenCommentNodes` call are being indiscriminately made ! Even though it seems like a micro optimization, perhaps we shoul consider walking only when corresponding types of checks are present (ordinary - javadoc). Please adress

**Comment by Vladlis** [comment]
@ps-sp 
Yes, make an issue pleasy, but don't waste your time until it is approved. The issue will require `checkstyle8` label and probably it will break a few checks.

> Also, walk calls and appendHiddenCommentNodes call are being indiscriminately made ! Even though it seems like a micro optimization

You are going off topic

**Comment by romani** [comment]
@ps-sp , @Vladlis , please focus on testing of this PR.

simple comparison for trees will be enough for this issue. There are should not too many cases in javadoc of opensource projects. In your case exception should be converted to tree, all other trees should not change - very easy.

Please make bash script or any other script to do this testing.
More detailed regression testing of AST trees is not easy discussion, might not be quick.

**Comment by ps-sp** [comment]
@romani got it. Already working on that.

**Comment by rnveach** [comment]
I am not finding any differences in the projects I ran this on:
http://rveach.no-ip.org/checkstyle/regression/reports/65/
http://rveach.no-ip.org/checkstyle/regression/reports/66/

**Comment by ps-sp** [comment]
@romani Please address. The diff reports have produced no difference.

**Comment by romani** [comment]
@ps-sp , 

1) please rebase on latest code to resolve conflicts.
2) please provide results of regression.

**Comment by rnveach** [comment]
@ps-sp Please also re-run regression with our beta ANTLR regression script found at https://github.com/checkstyle/contribution/issues/237#issuecomment-307267485 and confirm the results I gave are still the same.

**Comment by rnveach** [comment]
@ps-sp ping

**Comment by ps-sp** [comment]
@rnveach Yeah I will get the regression done ASAP.

**Comment by ps-sp** [comment]
[diff report1](https://issue-4164-diffreports1.netlify.com/)

**Comment by ps-sp** [comment]
@romani @rnveach @Vladlis Why don't we list down all the projects that must be present in the diff report for javadoc regression no matter what. Also, please do check the diff reports that I have uploaded.
